src/cockpit/channels/filesystem.py

```diff
@@ -160,6 +161,51 @@ def do_yield_data(self, options: JsonObject) -> Generator[bytes, None, JsonObjec
             raise ChannelError('internal-error', message=str(exc)) from exc
 
 
+class FSReplaceAttrs:
+    uid: 'int | None' = None
+    gid: 'int | None' = None
+    supported_attrs = ['owner', 'group']
+
+    def __init__(self, value: JsonObject) -> None:
+        # Any unknown keys throw an error
+        unsupported_attrs = list(value.keys() - self.supported_attrs)
+        if unsupported_attrs:
+            raise ChannelError('protocol-error',
+                               message=f'"attrs" contains unsupported key(s) {unsupported_attrs}',
+                               unsupported_attrs=unsupported_attrs) from None
+
+        try:
+            owner = get_str_or_int(value, 'owner', None)
+        except JsonError:
+            raise ChannelError('protocol-error', message='"owner" must be an integer or string') from None
+
+        try:
+            group = get_str_or_int(value, 'group', None)
+        except JsonError:
+            raise ChannelError('protocol-error', message='"group" must be an integer or string') from None
+
+        if owner is not None and group is None:
+            raise ChannelError('protocol-error', message='"group" attribute is empty while "owner" is provided')
+        if group is not None and owner is None:
+            raise ChannelError('protocol-error', message='"owner" attribute is empty while "group" is provided')
```

Comment:
This seems arbitrary and also seems like it makes the implementation more complicated, not less...

Reply:
Howso? @martinpitt wanted to make it required, you don't? Please discuss this. It is really not that much more complicated, costs 4 lines.

Reply:
My other rationale is that it's easier in the future to relax this restriction (that's API compatible) than to introduce it (as that's an API break)

Reply:
> I could live with setting the user without an explicit group if it then defaults to also setting the group to the primary user's group.

I think I disagree with this.  I'd want the same semantics as when you pass `-1` to `chown()`: it preserves the default.  Also: in the case of numeric uid we might not even have a primary group...

I think we're slowly iterating towards "forbid it until we can agree on how it would work".

Reply:
An explicit `group: -1` would be better, yes. But TBH I wouldn't add that until we actually have a demand for that or someone asks for it. Just for clarity, I don't *like* setting the user without an explicit group. I just thought you wanted that. So I'm completely on board with the current check and disallowing it outright.

Reply:
As discussed in our meeting today we'll keep both options required